### PR TITLE
optimize time window

### PR DESCRIPTION
**Pull request by iamlinjunhong**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #18244 

## What this PR does / why we need it:

optimize time window
1. add agg operator before order by
2. split avg for agg and time window
3. remove primary key constraints

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Optimized and enhanced the time window processing logic, including support for sliding intervals.
- Introduced new aggregation functions `AvgTwCache` and `AvgTwResult` for time window operations.
- Enhanced query builder and binders to support new time window expressions and functions.
- Added new built-in functions `Truncate` and `Divisor` for interval calculations.
- Updated tests to cover new time window logic and functions.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>14 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>timewin.go</strong><dd><code>Optimize and enhance time window processing logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/colexec/timewin/timewin.go

<li>Added new logic for handling time windows with sliding intervals.<br> <li> Refactored the <code>Prepare</code> and <code>Call</code> methods for better performance.<br> <li> Introduced new helper functions for timestamp expressions.<br> <li> Removed redundant code and improved error handling.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-c6bf0e244979295170141b3204fb3beb8c5a9320b6c2ae06d50908d8c8a3590c">+325/-513</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>query_builder.go</strong><dd><code>Enhance query builder for time window support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/query_builder.go

<li>Enhanced time window query building with new expressions.<br> <li> Added support for sliding intervals in time windows.<br> <li> Improved column reference remapping for time windows.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-144e02a38da50867dc021b9254d10a5fd131671cca344c2323337ba2141440f1">+148/-39</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>avg_tw_cache.go</strong><dd><code>Add AvgTwCache aggregation functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/function/agg/avg_tw_cache.go

<li>Introduced <code>AvgTwCache</code> aggregation functions.<br> <li> Registered new aggregation functions for various data types.<br> <li> Implemented context and merge logic for <code>AvgTwCache</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-609bd977e498ae68b1e71058bc83fe921f8fcc40786d7fea2659d6259b3dab46">+294/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>avg_tw_result.go</strong><dd><code>Introduce AvgTwResult aggregation functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/function/agg/avg_tw_result.go

<li>Added <code>AvgTwResult</code> aggregation functions.<br> <li> Implemented context and merge logic for <code>AvgTwResult</code>.<br> <li> Registered new aggregation functions for result types.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-a605f37d9ea93cb3eeadb6569d3400205d362bc62cf1c412f3fa514654f1c560">+212/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>types.go</strong><dd><code>Update time window types and logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/colexec/timewin/types.go

<li>Updated container and time window types for new logic.<br> <li> Added methods for interval and sliding calculations.<br> <li> Improved reset and release logic for time windows.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-d289628b7a7184e5c75212b430b5f890055de8dbb6add41d33ba118df962112a">+110/-81</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>having_binder.go</strong><dd><code>Enhance having binder for time window functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/having_binder.go

<li>Enhanced binding logic for time window functions.<br> <li> Added support for remapping aggregate functions in time windows.<br> <li> Improved handling of sliding intervals in having binder.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-f9b58dd9ed4aab99e4298ef1add98f7d2e3704a5aba924e2d9e13720b7b2f762">+89/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>func_binary.go</strong><dd><code>Add Truncate and Divisor functions for intervals</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/function/func_binary.go

<li>Added new functions <code>Truncate</code> and <code>Divisor</code>.<br> <li> Implemented logic for interval calculations.<br> <li> Enhanced error handling for unsupported units.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-730e142860376b56a5973dece41bb2643a9bc3b749ff5c140540ad3ee78a8e09">+102/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>base_binder.go</strong><dd><code>Enhance base binder for time window functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/base_binder.go

<li>Enhanced function binding logic for time window support.<br> <li> Added handling for new interval functions.<br> <li> Improved error handling and casting logic.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-d2c4ebed110cf4c8982fa501ff026b2d518810445a72f51eac3021b51396e7b0">+103/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>operator.go</strong><dd><code>Update time window construction logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/compile/operator.go

<li>Updated time window construction logic.<br> <li> Integrated process handling in time window construction.<br> <li> Improved error handling and argument setup.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-24c2df3f5c8c484aab6845aa35e6426ba672758aea5c19bb680c08eadec260ee">+6/-23</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>function_id.go</strong><dd><code>Add new function IDs for time window functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/function/function_id.go

<li>Added new function IDs for <code>AVG_TW_CACHE</code> and <code>AVG_TW_RESULT</code>.<br> <li> Registered new function IDs for <code>TRUNCATE</code> and <code>DIVISOR</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-8a340cc7bfd15d211dc0f230d234ba4edded44e79655671a94ce13f5fabfc7de">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>projection_binder.go</strong><dd><code>Enhance projection binder for time window expressions</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/projection_binder.go

<li>Enhanced projection binder for time window expressions.<br> <li> Added handling for time window start and end expressions.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-85b22445078685aeafccb7377098d93ebf0536af494ae2bf9c79ec9014952191">+15/-15</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>list_builtIn.go</strong><dd><code>Add built-in functions for interval operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/function/list_builtIn.go

<li>Added built-in functions <code>TRUNCATE</code> and <code>DIVISOR</code>.<br> <li> Registered new overloads for interval functions.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-b9b859604bf3c01505320f0a72c9862667f39cdf1323a88d7f8b2991d32098b8">+46/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>list_agg.go</strong><dd><code>Add aggregation functions for time window support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/function/list_agg.go

<li>Added aggregation functions <code>AVG_TW_CACHE</code> and <code>AVG_TW_RESULT</code>.<br> <li> Registered new aggregation logic for time window functions.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-28b8c798a448a1a44bb9deb400c059ce8ca0a5b0cae0e9e1ee884b21ef14b619">+42/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>fill.go</strong><dd><code>Update fill logic for new aggregation functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/colexec/fill/fill.go

<li>Updated fill logic to support new aggregation functions.<br> <li> Added handling for <code>AVG_TW_RESULT</code> in fill operations.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-aa15abed44aa6d7f944efdd2c2a090bdae777f103e8bcfcc281e8cd297e688dd">+5/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>timewin_test.go</strong><dd><code>Update tests for time window enhancements</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/colexec/timewin/timewin_test.go

<li>Updated tests for new time window logic.<br> <li> Added test cases for interval and sliding calculations.<br> <li> Improved test setup and validation.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18863/files#diff-f20e86ba88f6f961e385a77ffaf4a26e807e72352dee1cbb52a95737ac12f8e0">+8/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by m-schen — APPROVED** [review]

File: pkg/container/types/datetime.go [hunk 1]
```diff
@@ -25,7 +25,10 @@ import (
 )
 
 const (
-	SecsPerMinute       = 60
+	SecsPerMinute = 60
+	MinutePerHour = 60
+	MinutePerDay  = 24 * MinutePerHour
+	HourPerMinute
```
Comment (fengttt) [inline pkg/container/types/datetime.go:31]: What is HourPerMinute ...

**Review by heni02 — APPROVED** [review]

**Review by badboynt1 — APPROVED** [review]

File: pkg/sql/colexec/fill/fill.go [hunk 2]
```diff
@@ -50,7 +50,11 @@ func (fill *Fill) Prepare(proc *process.Process) (err error) {
 
 	f := true
 	for i := len(fill.AggIds) - 1; i >= 0; i-- {
-		if fill.AggIds[i] == function.MAX || fill.AggIds[i] == function.MIN || fill.AggIds[i] == function.SUM || fill.AggIds[i] == function.AVG {
+		if fill.AggIds[i] == function.MAX ||
```
Comment (fengttt) [inline pkg/sql/colexec/fill/fill.go:53]: Overall this is bad code -- should never test fill.AggIds == MIN etc.    Should always be from aggIds, get a function object, then check the property of the function.    

Basically, if I need to add a new aggregate function, all I should do is to add function def/impl in the function package, never need to change code here.

**Review by XuPeng-SH — APPROVED** [review]

File: test/distributed/cases/window/time_window.result [hunk 3]
```diff
@@ -349,6 +353,9 @@ select _wstart, _wend, sum(col2) from sliding_window01 where ts > "2023-08-01 00
 _wstart    _wend    sum(col2)
 2023-08-01 00:00:00    2023-08-01 01:00:00    177.0
 2023-08-01 00:10:00    2023-08-01 01:10:00    151.0
+2023-08-01 00:20:00    2023-08-01 01:20:00    123.0
```
Comment (fengttt) [inline test/distributed/cases/window/time_window.result:356]: Can you please double check if the result is correct?   Is this change saying that old result was WRONG?

File: pkg/sql/plan/function/function_id.go [hunk 4]
```diff
@@ -116,6 +118,7 @@ const (
 	TIMEDIFF
 	TIMESTAMPDIFF
 	DENSE_RANK
+	DIVISOR
```
Comment (fengttt) [inline pkg/sql/plan/function/function_id.go:121]: Are truncate and divisor well known SQL functions?   If only used in window, or internally, we should give them a better name, maybe prefix MO_WIN or something.

**Review by zhangxu19830126 — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

File: pkg/container/types/datetime.go (hunk 1 above)
Comment (iamlinjunhong) [reply pkg/container/types/datetime.go:31]: Not used, will be deleted

File: test/distributed/cases/window/time_window.result (hunk 3 above)
Comment (iamlinjunhong) [reply test/distributed/cases/window/time_window.result:356]: The old result is incorrect, there was a bug in the units of sliding and interval that were different before

File: pkg/sql/plan/function/function_id.go (hunk 4 above)
Comment (iamlinjunhong) [reply pkg/sql/plan/function/function_id.go]: done

File: pkg/sql/colexec/fill/fill.go (hunk 2 above)
Comment (iamlinjunhong) [reply pkg/sql/colexec/fill/fill.go]: bad code, just delete. I forgot to delete it before

**Review by ouyuanning — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Pull request merged**